content/blog/2025-12-16-bracket.md

```diff
@@ -0,0 +1,261 @@
+
+This representation enables efficient operations: for example, `(* 8 n)` directly represents fixnum `n`, allowing addition and subtraction without untagging since `8x + 8y = 8(x + y)`.
+
+Since Bril's core instruction set doesn't include bitwise operations like `bitwise-and`, `bitwise-ior`, `bitwise-xor`, and `arithmetic-shift-right`, I used GenAI to generate a **standard library** that implements these operations using only Bril's integer arithmetic and boolean logic:
+
+- **`@bit(x: int): int`** - Extracts the least significant bit
+- **`@bitand(a: int, b: int): int`** - Implements bitwise AND through loops
+- **`@bitor(a: int, b: int): int`** - Implements bitwise OR
+- **`@bitxor(a: int, b: int): int`** - Implements bitwise XOR
+- **`@ashr(x: int, n: int): int`** - Implements arithmetic shift right with sign extension
+
+I implemented the logic in `bracket` to recognize when a `binop` is a bitwise operation and generate the appropriate calls to these standard library functions. These functions are prepended to every compiled program.
+
+#### Control Flow Translation
+
+The source IR uses predicates (relops, boolean combinators) that must be converted to Bril's branch instructions. My `compile-pred` function:
+- Translates relops directly to Bril comparison operations
+- Handles `!=` by compiling as `(not (= op1 op2))`
+- Transforms `not`, `(true)`, and `(false)` appropriately
+- Threads effects through predicate computation
+- Handles nested `if` expressions by creating fresh labels
+
+#### Function Calls and Tail Calls
+
+The compiler distinguishes between tail calls and non-tail calls:
+- Tail calls in `main` untag the result (divide by 8) before printing
+- Tail calls in other functions emit a `ret` instruction
+- Non-tail calls store results in fresh temporaries
+- All arguments are evaluated to temporaries before calling
+
+#### Module Structure
+
+The source program consists of a module with multiple function definitions plus a main entry expression. I compile this by:
+- Creating a Bril function for each defined function with `any`-typed parameters
+- Creating a special `main` function that compiles the module's entry expression
+- Prepending the standard library functions to the output JSON
+
+### Technical Details
+
+The compiler uses helper functions for:
+- Converting abstract locations and labels to Bril variable/function names
+- Generating unique temporary variable and label names
+- Creating Bril instruction JSON structures
+- Handling operands (integers need const instructions, variables are used directly)
+- Accumulating instructions through emit closures for straightforward linear translation
+
+## What were the hardest parts to get right?
+
+### Choosing the Right Intermediate Language
+
+Selecting `proc-imp-cmf-lang-v7` required careful consideration. Lower-level languages would have required building an abstract interpreter for labels-as-values and physical-to-abstract location mapping.
```

Comment:
You can probably collapse this into a single section heading, or just delete it altogether, since you've already covered this well elsewhere.